File: backtrack/factor_combinations/factor_combinations.ex
defmodule Solution do
  def divisors(n, i, up_to) do
    if up_to < i do
      []
    else
      i..up_to |> Enum.filter(fn k -> rem(n, k) === 0 end)
    end
  end

  def getFactors(n, i \\ 2) do
    up_to = Kernel.trunc(:math.floor(:math.sqrt(n)))
    divisors(n, i, up_to)
    |> Enum.flat_map(
        fn d ->
          [[div(n, d)] | getFactors(div(n,d), d)]
            |> Enum.map(fn s -> [d] ++ s  end)
        end
      )
  end
end


IO.inspect Solution.getFactors(16)

File: backtrack/factor_combinations/factor_combinations.scala
def getFactors(n: Int, i: Int = 2): List[List[Int]] = {
  val up_to = math.floor(math.sqrt(n)).toInt
  val divisors = (for (k <- i to up_to if n % k == 0) yield k).toList
  // calculate al decompositions starting with d
  divisors.flatMap(d => {
    // add n/d to the list of remaining factors
    (List(n / d) :: getFactors(n / d, d))
      // add d to the head of each subcombination
      .map(s => d :: s)
  }
  )
}


println(getFactors(16))
